src/frontend/src/expr/utils.rs

```diff
@@ -587,12 +583,174 @@ impl WatermarkAnalyzer {
                 // TODO: do we need derive watermark when every case can derive a common watermark?
                 WatermarkDerivation::None
             }
-            ExprType::Proctime => Nondecreasing,
+            ExprType::Proctime => WatermarkDerivation::None,
             _ => WatermarkDerivation::None,
         }
     }
 }
 
+#[derive(Clone, Default)]
+pub struct MonotonicityAnalyzer {
+    pub input_mono_cols: Vec<(usize, Monotonicity)>,
```

Comment:
What about storing input ref here? So that we don't need construct the `Vec`, and also we are able to take advantage of other properties of input if needed in the future.